src/apps/vendor-finance/common/vendor-finance.pool-v2.contract-position-fetcher.ts

```diff
@@ -0,0 +1,166 @@
+      { metaType: MetaType.SUPPLIED, address: definition.lendToken, network: this.network },
+    ];
+  }
+
+  async getLabel({
+    contractPosition,
+    definition,
+  }: GetDisplayPropsParams<VendorFinancePoolV2, DefaultDataProps, VendorFinancePoolV2Definition>) {
+    const poolLabel = `${getLabelFromToken(contractPosition.tokens[0])}/${getLabelFromToken(
+      contractPosition.tokens[1],
+    )}`;
+
+    const dateString = new Date(parseInt(definition.expiry) * 1000).toLocaleDateString('en-US', {
+      year: 'numeric',
+      month: 'long',
+      day: 'numeric',
+      hour: '2-digit',
+      minute: '2-digit',
+    });
+
+    return `${poolLabel} Lending Pool - ${dateString}`;
+  }
+
+  async getStatsItems({
+    contractPosition,
+    definition,
+  }: GetDisplayPropsParams<VendorFinancePoolV2, DefaultDataProps, VendorFinancePoolV2Definition>) {
+    const debtToken = contractPosition.tokens.find(isBorrowed)!;
+
+    return [
+      {
+        label: 'Lending APR',
+        value: buildPercentageDisplayItem(parseInt(definition.feeRate) / 10000),
+      },
+      {
+        label: 'Repayment date',
+        value: new Date(parseInt(definition.expiry) * 1000).toDateString(),
+      },
+      {
+        label: 'Available',
+        value: buildDollarDisplayItem((parseInt(definition.lendBalance) / 10 ** debtToken.decimals) * debtToken.price),
+      },
+      {
+        label: 'Borrowed',
+        value: buildDollarDisplayItem(
+          (parseInt(definition.totalBorrowed) / 10 ** debtToken.decimals) * debtToken.price,
+        ),
+      },
+      {
+        label: 'Lend ratio',
+        value: parseInt(definition.mintRatio) / 10 ** 18,
+      },
+    ];
+  }
+  async getDataProps(
+    _params: GetDataPropsParams<VendorFinancePoolV2, DefaultDataProps, VendorFinancePoolV2Definition>,
+  ): Promise<VendorFinancePoolDataProps> {
+    return {
+      deployer: _params.definition.deployer,
+      totalDeposited: parseInt(_params.definition.lendBalance) + parseInt(_params.definition.totalBorrowed),
+    };
+  }
+
+  async getTokenBalancesPerPosition({
+    address,
+    contractPosition,
+  }: GetTokenBalancesParams<VendorFinancePoolV2, VendorFinancePoolDataProps>) {
+    const collateralToken = contractPosition.tokens[0]!;
+    const lentToken = contractPosition.tokens[1]!;
+
+    // --- Lender logic ----
+    // No deposit, no borrow, but lending out
+    if (address === contractPosition.dataProps.deployer.toLowerCase()) {
+      return ['0', '0', contractPosition.dataProps.totalDeposited.toString()];
+    }
+    // --! Lender logic !---
+
+    // --- Borrower logic ----
+    const data = await gqlFetch<VendorBorrowerV2GraphResponse>({
+      endpoint: this.subgraphUrl,
+      query: borrowerV2InfosQuery(address),
+    });
+
+    const borrowerPosition = data.borrower?.positions.find(({ pool }) => pool.id === contractPosition.address);
+    if (!borrowerPosition) return [];
```

Comment:
I'm wondering if it would be possible to fetch the balance calling the chain directly and not rely on the subgraph for `the balance`. I saw the `pool` position is also fetching the balance using the subgraph, but it's def something we want to move away from.

Reply:
I have eliminated the requirement to call the subgraph to fetch borrow positions and their corresponding balances. In the case of v2 pools, there is an on-chain position tracker smart contract available. This contract keeps track of the pools in which a user has positions, but does not track balances directly. To determine a user's debt balance, the position tracker contract can be queried to identify the relevant pool(s), and then each pool must be called to retrieve the debt balance associated with the user.

The setup in my latest commits can be summarized as follows:

1) Retrieve an instance of the position tracker contract.
2) Obtain the borrow positions associated with a specified address.
3) Fetch an instance of each lending pool contract in which the user has a borrow position. This step is necessary because the actual debt balance of a borrower can only be found in a specific `mapping` within each pool.
4) Accumulate the debt from all relevant pools into a variable called `totalBorrowed`. This variable represents the sum of the borrowed amounts. 

The key difference between using the subgraph and adopting this latest approach lies in the accessibility of the `totalBorrowed` variable. In the case of the subgraph, the `totalBorrowed` value can be directly retrieved, whereas when excluding the subgraph, an alternative method is employed to derive the `totalBorrowed` amount.